internal/contextualtuples/contextual_tuples.go

```diff
@@ -0,0 +1,66 @@
+package contextualtuples
+
+import (
+	"github.com/openfga/openfga/internal/validation"
+	"github.com/openfga/openfga/pkg/typesystem"
+
+	tupleUtils "github.com/openfga/openfga/pkg/tuple"
+	serverErrors "github.com/openfga/openfga/server/errors"
+	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
+)
+
+type ContextualTuples struct {
+	// map where key is of form 'object#relation' and value is the list of tuples that are of form 'object#relation@X'
+	usersets map[string][]*openfgapb.TupleKey
+}
+
+// New validates the input tuples against the typesystem and, if valid,
+// returns those tuples as ContextualTuples
+func New(typesys *typesystem.TypeSystem, tupleKeys []*openfgapb.TupleKey) (*ContextualTuples, error) {
+	tupleMap := map[string]struct{}{}
+	usersets := map[string][]*openfgapb.TupleKey{}
+	for _, tk := range tupleKeys {
+		if _, ok := tupleMap[tk.String()]; ok {
+			return nil, serverErrors.DuplicateContextualTuple(tk)
+		}
+		tupleMap[tk.String()] = struct{}{}
+
+		if err := validation.ValidateTuple(typesys, tk); err != nil {
+			return nil, serverErrors.HandleTupleValidateError(err)
+		}
+
+		key := tupleUtils.ToObjectRelationString(tk.GetObject(), tk.GetRelation())
+		usersets[key] = append(usersets[key], tk)
+	}
+
+	return &ContextualTuples{usersets: usersets}, nil
+}
+
+// Read returns the list of contextual tuples that match 'object#relation'
+func (c *ContextualTuples) Read(tk *openfgapb.TupleKey) []*openfgapb.TupleKey {
+	return c.usersets[tupleUtils.ToObjectRelationString(tk.GetObject(), tk.GetRelation())]
+}
+
+// ReadUserTuple returns, if found, a contextual tuple that matches the input 'object#relation@user' and a bool indicating
+// whether such tuple was found.
+func (c *ContextualTuples) ReadUserTuple(tk *openfgapb.TupleKey) (*openfgapb.TupleKey, bool) {
```

Comment:
Wondering why we need to return the boolean.  Could the caller just check whether *openfgapb.TupleKey is nil?  I see that we just move this function from server/commands to here ... so, this should not block to merge.

Reply:
Yeah i moved stuff around. We can change this later. But, i worry whether we will forget to check for `nil` 🤔